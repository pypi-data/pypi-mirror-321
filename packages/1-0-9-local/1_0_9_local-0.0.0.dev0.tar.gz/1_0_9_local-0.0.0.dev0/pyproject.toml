[build-system]
build-backend = "setuptools.build_meta"
requires = [
    "setuptools>=61.0",
]

[project]
authors = [
    { email = "johannes-programming@mailfence.com", name = "Johannes" },
]
classifiers = [
    "Development Status :: 1 - Planning",
    "License :: OSI Approved :: MIT License",
    "Natural Language :: English",
    "Operating System :: OS Independent",
    "Programming Language :: Python",
    "Programming Language :: Python :: 3",
    "Programming Language :: Python :: 3 :: Only",
]
dependencies = []
description = "1_0_9_local"
keywords = []
name = "1_0_9_local"
readme = "README.rst"
requires-python = ">=3.12.7"
version = "0.0.0.dev0"

[project.license]
file = "LICENSE.txt"

[project.urls]
Download = "https://pypi.org/project/1_0_9_local/#files"
Index = "https://pypi.org/project/1_0_9_local/"
Source = "https://github.com/johannes-programming/1_0_9_local/"
