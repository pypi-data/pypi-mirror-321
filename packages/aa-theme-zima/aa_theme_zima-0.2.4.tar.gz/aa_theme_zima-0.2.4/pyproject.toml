[tool.poetry]
name = "aa-theme-zima"
version = "0.2.4"
description = "Theme for zimacorp.space"
authors = ["Boris Talovikov <boris@talovikov.ru>"]
license = "GPL-3.0-or-later"
readme = "README.md"
repository = "https://gitlab.com/zima-corp/aa-theme-zima"
packages = [{ include = "aa_theme_zima" }]

[tool.poetry.dependencies]
python = ">=3.8,<4.0"
allianceauth = ">=3.6"


[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"


[tool.pyright]
venv = ".venv"
venvPath = "."
