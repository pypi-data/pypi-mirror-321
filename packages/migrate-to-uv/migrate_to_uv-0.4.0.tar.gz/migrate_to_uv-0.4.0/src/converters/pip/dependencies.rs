use pep508_rs::Requirement;
use std::fs;
use std::path::Path;
use std::str::FromStr;
use url::Url;

pub fn get(project_path: &Path, requirements_files: Vec<String>) -> Option<Vec<String>> {
    let mut dependencies: Vec<String> = Vec::new();

    for requirements_file in requirements_files {
        let requirements_content =
            fs::read_to_string(project_path.join(requirements_file)).unwrap();

        for line in requirements_content.lines() {
            let dependency_specification = Requirement::<Url>::from_str(line);

            if let Ok(dependency_specification) = dependency_specification {
                dependencies.push(dependency_specification.to_string());
            }
        }
    }

    if dependencies.is_empty() {
        return None;
    }
    Some(dependencies)
}

pub fn get_constraint_dependencies(
    ignore_locked_versions: bool,
    is_pip_tools: bool,
    project_path: &Path,
    requirements_files: Vec<String>,
    dev_requirements_files: Vec<String>,
) -> Option<Vec<String>> {
    if !is_pip_tools || ignore_locked_versions {
        return None;
    }

    if let Some(dependencies) = get(
        project_path,
        requirements_files
            .into_iter()
            .chain(dev_requirements_files)
            .map(|f| f.replace(".in", ".txt"))
            .collect(),
    ) {
        if dependencies.is_empty() {
            return None;
        }
        return Some(dependencies);
    }
    None
}
