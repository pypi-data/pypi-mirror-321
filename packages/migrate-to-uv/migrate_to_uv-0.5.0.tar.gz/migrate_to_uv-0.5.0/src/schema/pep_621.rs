use crate::schema::poetry::Poetry;
use crate::schema::uv::Uv;
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

/// <https://packaging.python.org/en/latest/specifications/pyproject-toml/#pyproject-toml-spec>
#[derive(Default, Deserialize, Serialize)]
pub struct Project {
    pub name: Option<String>,
    pub version: Option<String>,
    pub description: Option<String>,
    pub authors: Option<Vec<AuthorOrMaintainer>>,
    #[serde(rename(serialize = "requires-python"))]
    pub requires_python: Option<String>,
    pub readme: Option<String>,
    pub license: Option<String>,
    pub maintainers: Option<Vec<AuthorOrMaintainer>>,
    pub keywords: Option<Vec<String>>,
    pub classifiers: Option<Vec<String>>,
    pub dependencies: Option<Vec<String>>,
    #[serde(rename(serialize = "optional-dependencies"))]
    pub optional_dependencies: Option<IndexMap<String, Vec<String>>>,
    pub urls: Option<IndexMap<String, String>>,
    pub scripts: Option<IndexMap<String, String>>,
    #[serde(rename(serialize = "gui-scripts"))]
    pub gui_scripts: Option<IndexMap<String, String>>,
    #[serde(rename(serialize = "entry-points"))]
    pub entry_points: Option<IndexMap<String, IndexMap<String, String>>>,
}

#[derive(Deserialize, Serialize)]
pub struct AuthorOrMaintainer {
    pub name: Option<String>,
    pub email: Option<String>,
}

#[derive(Deserialize, Serialize)]
pub struct Tool {
    pub poetry: Option<Poetry>,
    pub uv: Option<Uv>,
}
