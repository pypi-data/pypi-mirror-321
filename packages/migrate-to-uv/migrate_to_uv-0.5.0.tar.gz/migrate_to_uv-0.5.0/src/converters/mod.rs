use crate::schema::pyproject::DependencyGroupSpecification;
use indexmap::IndexMap;
use log::{error, info, warn};
use owo_colors::OwoColorize;
#[cfg(test)]
use std::any::Any;
use std::fmt::Debug;
use std::io::ErrorKind;
use std::path::Path;
use std::process::{Command, Stdio};

pub mod pip;
pub mod pipenv;
pub mod poetry;
mod pyproject_updater;

type DependencyGroupsAndDefaultGroups = (
    Option<IndexMap<String, Vec<DependencyGroupSpecification>>>,
    Option<Vec<String>>,
);

/// Converts a project from a package manager to uv.
pub trait Converter: Debug {
    #[allow(clippy::fn_params_excessive_bools)]
    fn convert_to_uv(
        &self,
        dry_run: bool,
        skip_lock: bool,
        ignore_locked_versions: bool,
        keep_old_metadata: bool,
        dependency_groups_strategy: DependencyGroupsStrategy,
    );

    #[cfg(test)]
    fn as_any(&self) -> &dyn Any;
}

pub fn lock_dependencies(project_path: &Path, is_removing_constraints: bool) -> Result<(), ()> {
    const UV_EXECUTABLE: &str = "uv";

    match Command::new(UV_EXECUTABLE)
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
    {
        Ok(_) => {
            info!(
                "Locking dependencies with \"{}\"{}...",
                format!("{UV_EXECUTABLE} lock").bold(),
                if is_removing_constraints {
                    " again to remove constraints"
                } else {
                    ""
                }
            );

            Command::new(UV_EXECUTABLE)
                .arg("lock")
                .current_dir(project_path)
                .spawn()
                .map_or_else(
                    |_| {
                        error!(
                            "Could not invoke \"{}\" command.",
                            format!("{UV_EXECUTABLE} lock").bold()
                        );
                        Err(())
                    },
                    |lock| match lock.wait_with_output() {
                        Ok(output) => {
                            if output.status.success() {
                                Ok(())
                            } else {
                                Err(())
                            }
                        }
                        Err(e) => {
                            error!("{e}");
                            Err(())
                        }
                    },
                )
        }
        Err(e) if e.kind() == ErrorKind::NotFound => {
            warn!(
                "Could not find \"{}\" executable, skipping locking dependencies.",
                UV_EXECUTABLE.bold()
            );
            Ok(())
        }
        Err(e) => {
            error!("{e}");
            Err(())
        }
    }
}

#[derive(clap::ValueEnum, Clone, Copy, Debug)]
pub enum DependencyGroupsStrategy {
    SetDefaultGroups,
    IncludeInDev,
    KeepExisting,
    MergeIntoDev,
}
