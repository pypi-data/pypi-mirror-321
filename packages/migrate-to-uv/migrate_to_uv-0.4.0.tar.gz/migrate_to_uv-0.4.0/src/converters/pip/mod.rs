mod dependencies;

use crate::converters::pip::dependencies::get_constraint_dependencies;
use crate::converters::pyproject_updater::PyprojectUpdater;
use crate::converters::DependencyGroupsStrategy;
use crate::converters::{lock_dependencies, Converter};
use crate::schema::pep_621::Project;
use crate::schema::pyproject::DependencyGroupSpecification;
use crate::schema::uv::Uv;
use crate::toml::PyprojectPrettyFormatter;
use indexmap::IndexMap;
use log::{info, warn};
use owo_colors::OwoColorize;
#[cfg(test)]
use std::any::Any;
use std::default::Default;
use std::fs;
use std::fs::{remove_file, File};
use std::io::Write;
use std::path::PathBuf;
use toml_edit::visit_mut::VisitMut;
use toml_edit::DocumentMut;

#[derive(Debug, PartialEq, Eq)]
pub struct Pip {
    pub project_path: PathBuf,
    pub requirements_files: Vec<String>,
    pub dev_requirements_files: Vec<String>,
    pub is_pip_tools: bool,
}

impl Converter for Pip {
    fn convert_to_uv(
        &self,
        dry_run: bool,
        skip_lock: bool,
        ignore_locked_versions: bool,
        keep_old_metadata: bool,
        _dependency_groups_strategy: DependencyGroupsStrategy,
    ) {
        let pyproject_path = self.project_path.join("pyproject.toml");
        let updated_pyproject_string = self.perform_migration(ignore_locked_versions);

        if dry_run {
            let mut pyproject_updater = PyprojectUpdater {
                pyproject: &mut updated_pyproject_string.parse::<DocumentMut>().unwrap(),
            };
            info!(
                "{}\n{}",
                "Migrated pyproject.toml:".bold(),
                pyproject_updater
                    .remove_constraint_dependencies()
                    .map_or(updated_pyproject_string, ToString::to_string)
            );
        } else {
            let mut pyproject_file = File::create(&pyproject_path).unwrap();

            pyproject_file
                .write_all(updated_pyproject_string.as_bytes())
                .unwrap();

            if !keep_old_metadata {
                self.delete_requirements_files().unwrap();
            }

            if !dry_run
                && !skip_lock
                && lock_dependencies(self.project_path.as_ref(), false).is_err()
            {
                warn!(
                    "An error occurred when locking dependencies, so \"{}\" was not created.",
                    "uv.lock".bold()
                );
            }

            // There are no locked dependencies for pip, so we only have to remove constraints for
            // pip-tools.
            if self.is_pip_tools && !ignore_locked_versions {
                let mut pyproject_updater = PyprojectUpdater {
                    pyproject: &mut updated_pyproject_string.parse::<DocumentMut>().unwrap(),
                };
                if let Some(updated_pyproject) = pyproject_updater.remove_constraint_dependencies()
                {
                    let mut pyproject_file = File::create(pyproject_path).unwrap();
                    pyproject_file
                        .write_all(updated_pyproject.to_string().as_bytes())
                        .unwrap();

                    // Lock dependencies a second time, to remove constraints from lock file.
                    if !dry_run
                        && !skip_lock
                        && lock_dependencies(self.project_path.as_ref(), true).is_err()
                    {
                        warn!("An error occurred when locking dependencies after removing constraints.");
                    }
                }
            }

            info!(
                "{}",
                format!(
                    "Successfully migrated project from {} to uv!\n",
                    if self.is_pip_tools {
                        "pip-tools"
                    } else {
                        "pip"
                    }
                )
                .bold()
                .green()
            );
        }
    }

    #[cfg(test)]
    fn as_any(&self) -> &dyn Any {
        self
    }
}

impl Pip {
    fn perform_migration(&self, ignore_locked_versions: bool) -> String {
        let dev_dependencies =
            dependencies::get(&self.project_path, self.dev_requirements_files.clone());

        let dependency_groups = dev_dependencies.map_or_else(
            || None,
            |dependencies| {
                let mut groups = IndexMap::new();

                groups.insert(
                    "dev".to_string(),
                    dependencies
                        .iter()
                        .map(|dep| DependencyGroupSpecification::String(dep.to_string()))
                        .collect(),
                );

                Some(groups)
            },
        );

        let project = Project {
            // "name" is required by uv.
            name: Some(String::new()),
            // "version" is required by uv.
            version: Some("0.0.1".to_string()),
            dependencies: dependencies::get(&self.project_path, self.requirements_files.clone()),
            ..Default::default()
        };

        let uv = Uv {
            package: Some(false),
            constraint_dependencies: get_constraint_dependencies(
                ignore_locked_versions,
                self.is_pip_tools,
                &self.project_path,
                self.requirements_files.clone(),
                self.dev_requirements_files.clone(),
            ),
            ..Default::default()
        };

        let pyproject_toml_content =
            fs::read_to_string(self.project_path.join("pyproject.toml")).unwrap_or_default();
        let mut updated_pyproject = pyproject_toml_content.parse::<DocumentMut>().unwrap();
        let mut pyproject_updater = PyprojectUpdater {
            pyproject: &mut updated_pyproject,
        };

        pyproject_updater.insert_pep_621(&project);
        pyproject_updater.insert_dependency_groups(dependency_groups.as_ref());
        pyproject_updater.insert_uv(&uv);

        let mut visitor = PyprojectPrettyFormatter {
            parent_keys: Vec::new(),
        };
        visitor.visit_document_mut(&mut updated_pyproject);

        updated_pyproject.to_string()
    }

    fn delete_requirements_files(&self) -> std::io::Result<()> {
        for requirements_file in self
            .requirements_files
            .iter()
            .chain(&self.dev_requirements_files)
        {
            let requirements_path = self.project_path.join(requirements_file);

            if requirements_path.exists() {
                remove_file(requirements_path.clone())?;
            }
        }

        // For pip-tools, also delete `.txt` files generated from `.in` files.
        if self.is_pip_tools {
            for requirements_file in self
                .requirements_files
                .iter()
                .chain(&self.dev_requirements_files)
                .map(|file| file.replace(".in", ".txt"))
            {
                let requirements_path = self.project_path.join(requirements_file);

                if requirements_path.exists() {
                    remove_file(requirements_path.clone())?;
                }
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_perform_pip_tools_migration() {
        let pip_tools = Pip {
            project_path: PathBuf::from("tests/fixtures/pip_tools/full"),
            requirements_files: vec!["requirements.in".to_string()],
            dev_requirements_files: vec!["requirements-dev.in".to_string()],
            is_pip_tools: true,
        };

        insta::assert_toml_snapshot!(pip_tools.perform_migration(true));
    }

    #[test]
    fn test_perform_pip_tools_all_files_migration() {
        let pip_tools = Pip {
            project_path: PathBuf::from("tests/fixtures/pip_tools/full"),
            requirements_files: vec!["requirements.in".to_string()],
            dev_requirements_files: vec![
                "requirements-dev.in".to_string(),
                "requirements-typing.in".to_string(),
            ],
            is_pip_tools: true,
        };

        insta::assert_toml_snapshot!(pip_tools.perform_migration(true));
    }

    #[test]
    fn test_perform_pip_tools_with_lock_file() {
        let pip_tools = Pip {
            project_path: PathBuf::from("tests/fixtures/pip_tools/with_lock_file"),
            requirements_files: vec!["requirements.in".to_string()],
            dev_requirements_files: vec![
                "requirements-dev.in".to_string(),
                "requirements-typing.in".to_string(),
            ],
            is_pip_tools: true,
        };

        insta::assert_toml_snapshot!(pip_tools.perform_migration(false));
    }

    #[test]
    fn test_perform_pip_migration() {
        let pip = Pip {
            project_path: PathBuf::from("tests/fixtures/pip/full"),
            requirements_files: vec!["requirements.txt".to_string()],
            dev_requirements_files: vec!["requirements-dev.txt".to_string()],
            is_pip_tools: false,
        };

        insta::assert_toml_snapshot!(pip.perform_migration(true));
    }

    #[test]
    fn test_perform_pip_all_files_migration() {
        let pip = Pip {
            project_path: PathBuf::from("tests/fixtures/pip/full"),
            requirements_files: vec!["requirements.txt".to_string()],
            dev_requirements_files: vec![
                "requirements-dev.txt".to_string(),
                "requirements-typing.txt".to_string(),
            ],
            is_pip_tools: false,
        };

        insta::assert_toml_snapshot!(pip.perform_migration(true));
    }
}
