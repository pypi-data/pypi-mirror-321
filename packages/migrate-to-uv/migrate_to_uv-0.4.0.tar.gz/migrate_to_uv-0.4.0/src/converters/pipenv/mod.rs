mod dependencies;
mod project;
mod sources;

use crate::converters::pipenv::dependencies::get_constraint_dependencies;
use crate::converters::pyproject_updater::PyprojectUpdater;
use crate::converters::DependencyGroupsStrategy;
use crate::converters::{lock_dependencies, Converter};
use crate::schema::pep_621::Project;
use crate::schema::pipenv::Pipfile;
use crate::schema::uv::{SourceContainer, Uv};
use crate::toml::PyprojectPrettyFormatter;
use indexmap::IndexMap;
use log::{info, warn};
use owo_colors::OwoColorize;
#[cfg(test)]
use std::any::Any;
use std::default::Default;
use std::fs;
use std::fs::{remove_file, File};
use std::io::Write;
use std::path::PathBuf;
use toml_edit::visit_mut::VisitMut;
use toml_edit::DocumentMut;

#[derive(Debug, PartialEq, Eq)]
pub struct Pipenv {
    pub project_path: PathBuf,
}

impl Converter for Pipenv {
    fn convert_to_uv(
        &self,
        dry_run: bool,
        skip_lock: bool,
        ignore_locked_versions: bool,
        keep_old_metadata: bool,
        dependency_groups_strategy: DependencyGroupsStrategy,
    ) {
        let pyproject_path = self.project_path.join("pyproject.toml");
        let updated_pyproject_string =
            self.perform_migration(ignore_locked_versions, dependency_groups_strategy);

        if dry_run {
            let mut pyproject_updater = PyprojectUpdater {
                pyproject: &mut updated_pyproject_string.parse::<DocumentMut>().unwrap(),
            };
            info!(
                "{}\n{}",
                "Migrated pyproject.toml:".bold(),
                pyproject_updater
                    .remove_constraint_dependencies()
                    .map_or(updated_pyproject_string, ToString::to_string)
            );
        } else {
            let mut pyproject_file = File::create(&pyproject_path).unwrap();

            pyproject_file
                .write_all(updated_pyproject_string.as_bytes())
                .unwrap();

            if !keep_old_metadata {
                self.delete_pipenv_references().unwrap();
            }

            if !dry_run
                && !skip_lock
                && lock_dependencies(self.project_path.as_ref(), false).is_err()
            {
                warn!(
                    "An error occurred when locking dependencies, so \"{}\" was not created.",
                    "uv.lock".bold()
                );
            }

            if !ignore_locked_versions {
                let mut pyproject_updater = PyprojectUpdater {
                    pyproject: &mut updated_pyproject_string.parse::<DocumentMut>().unwrap(),
                };
                if let Some(updated_pyproject) = pyproject_updater.remove_constraint_dependencies()
                {
                    let mut pyproject_file = File::create(pyproject_path).unwrap();
                    pyproject_file
                        .write_all(updated_pyproject.to_string().as_bytes())
                        .unwrap();

                    // Lock dependencies a second time, to remove constraints from lock file.
                    if !dry_run
                        && !skip_lock
                        && lock_dependencies(self.project_path.as_ref(), true).is_err()
                    {
                        warn!("An error occurred when locking dependencies after removing constraints.");
                    }
                }
            }

            info!(
                "{}",
                "Successfully migrated project from Pipenv to uv!\n"
                    .bold()
                    .green()
            );
        }
    }

    #[cfg(test)]
    fn as_any(&self) -> &dyn Any {
        self
    }
}

impl Pipenv {
    fn perform_migration(
        &self,
        ignore_locked_versions: bool,
        dependency_groups_strategy: DependencyGroupsStrategy,
    ) -> String {
        let pipfile_content = fs::read_to_string(self.project_path.join("Pipfile")).unwrap();
        let pipfile: Pipfile = toml::from_str(pipfile_content.as_str()).unwrap();

        let mut uv_source_index: IndexMap<String, SourceContainer> = IndexMap::new();
        let (dependency_groups, uv_default_groups) =
            dependencies::get_dependency_groups_and_default_groups(
                &pipfile,
                &mut uv_source_index,
                dependency_groups_strategy,
            );

        let project = Project {
            // "name" is required by uv.
            name: Some(String::new()),
            // "version" is required by uv.
            version: Some("0.0.1".to_string()),
            requires_python: project::get_requires_python(pipfile.requires),
            dependencies: dependencies::get(pipfile.packages.as_ref(), &mut uv_source_index),
            ..Default::default()
        };

        let uv = Uv {
            package: Some(false),
            index: sources::get_indexes(pipfile.source),
            sources: if uv_source_index.is_empty() {
                None
            } else {
                Some(uv_source_index)
            },
            default_groups: uv_default_groups,
            constraint_dependencies: get_constraint_dependencies(
                ignore_locked_versions,
                &self.project_path.join("Pipfile.lock"),
            ),
        };

        let pyproject_toml_content =
            fs::read_to_string(self.project_path.join("pyproject.toml")).unwrap_or_default();
        let mut updated_pyproject = pyproject_toml_content.parse::<DocumentMut>().unwrap();
        let mut pyproject_updater = PyprojectUpdater {
            pyproject: &mut updated_pyproject,
        };

        pyproject_updater.insert_pep_621(&project);
        pyproject_updater.insert_dependency_groups(dependency_groups.as_ref());
        pyproject_updater.insert_uv(&uv);

        let mut visitor = PyprojectPrettyFormatter {
            parent_keys: Vec::new(),
        };
        visitor.visit_document_mut(&mut updated_pyproject);

        updated_pyproject.to_string()
    }

    fn delete_pipenv_references(&self) -> std::io::Result<()> {
        let pipfile_path = self.project_path.join("Pipfile");

        if pipfile_path.exists() {
            remove_file(pipfile_path)?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_perform_migration() {
        let pipenv = Pipenv {
            project_path: PathBuf::from("tests/fixtures/pipenv/full"),
        };

        insta::assert_toml_snapshot!(
            pipenv.perform_migration(true, DependencyGroupsStrategy::SetDefaultGroups)
        );
    }

    #[test]
    fn test_perform_migration_dep_group_include_in_dev() {
        let pipenv = Pipenv {
            project_path: PathBuf::from("tests/fixtures/pipenv/full"),
        };

        insta::assert_toml_snapshot!(
            pipenv.perform_migration(true, DependencyGroupsStrategy::IncludeInDev)
        );
    }

    #[test]
    fn test_perform_migration_dep_group_keep_existing() {
        let pipenv = Pipenv {
            project_path: PathBuf::from("tests/fixtures/pipenv/full"),
        };

        insta::assert_toml_snapshot!(
            pipenv.perform_migration(true, DependencyGroupsStrategy::KeepExisting)
        );
    }

    #[test]
    fn test_perform_migration_dep_group_merge_in_dev() {
        let pipenv = Pipenv {
            project_path: PathBuf::from("tests/fixtures/pipenv/full"),
        };

        insta::assert_toml_snapshot!(
            pipenv.perform_migration(true, DependencyGroupsStrategy::MergeIntoDev)
        );
    }

    #[test]
    fn test_perform_migration_python_full_version() {
        let pipenv = Pipenv {
            project_path: PathBuf::from("tests/fixtures/pipenv/python_full_version"),
        };

        insta::assert_toml_snapshot!(
            pipenv.perform_migration(true, DependencyGroupsStrategy::SetDefaultGroups)
        );
    }

    #[test]
    fn test_perform_migration_empty_requires() {
        let pipenv = Pipenv {
            project_path: PathBuf::from("tests/fixtures/pipenv/empty_requires"),
        };

        insta::assert_toml_snapshot!(
            pipenv.perform_migration(true, DependencyGroupsStrategy::SetDefaultGroups)
        );
    }

    #[test]
    fn test_perform_migration_minimal_pipfile() {
        let pipenv = Pipenv {
            project_path: PathBuf::from("tests/fixtures/pipenv/minimal"),
        };

        insta::assert_toml_snapshot!(
            pipenv.perform_migration(true, DependencyGroupsStrategy::SetDefaultGroups)
        );
    }

    #[test]
    fn test_perform_migration_with_lock_file() {
        let pipenv = Pipenv {
            project_path: PathBuf::from("tests/fixtures/pipenv/with_lock_file"),
        };

        insta::assert_toml_snapshot!(
            pipenv.perform_migration(false, DependencyGroupsStrategy::SetDefaultGroups)
        );
    }
}
