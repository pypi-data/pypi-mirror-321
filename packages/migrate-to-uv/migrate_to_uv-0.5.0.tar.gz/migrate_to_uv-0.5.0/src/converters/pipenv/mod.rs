mod dependencies;
mod project;
mod sources;

use crate::converters::pipenv::dependencies::get_constraint_dependencies;
use crate::converters::pyproject_updater::PyprojectUpdater;
use crate::converters::DependencyGroupsStrategy;
use crate::converters::{lock_dependencies, Converter};
use crate::schema::pep_621::Project;
use crate::schema::pipenv::Pipfile;
use crate::schema::uv::{SourceContainer, Uv};
use crate::toml::PyprojectPrettyFormatter;
use indexmap::IndexMap;
use log::{info, warn};
use owo_colors::OwoColorize;
#[cfg(test)]
use std::any::Any;
use std::default::Default;
use std::fs;
use std::fs::{remove_file, File};
use std::io::Write;
use std::path::PathBuf;
use toml_edit::visit_mut::VisitMut;
use toml_edit::DocumentMut;

const FILES_TO_DELETE: &[&str] = &["Pipfile", "Pipfile.lock"];

#[derive(Debug, PartialEq, Eq)]
pub struct Pipenv {
    pub project_path: PathBuf,
}

impl Converter for Pipenv {
    fn convert_to_uv(
        &self,
        dry_run: bool,
        skip_lock: bool,
        ignore_locked_versions: bool,
        keep_old_metadata: bool,
        dependency_groups_strategy: DependencyGroupsStrategy,
    ) {
        let pyproject_path = self.project_path.join("pyproject.toml");
        let updated_pyproject_string =
            self.perform_migration(ignore_locked_versions, dependency_groups_strategy);

        if dry_run {
            let mut pyproject_updater = PyprojectUpdater {
                pyproject: &mut updated_pyproject_string.parse::<DocumentMut>().unwrap(),
            };
            info!(
                "{}\n{}",
                "Migrated pyproject.toml:".bold(),
                pyproject_updater
                    .remove_constraint_dependencies()
                    .map_or(updated_pyproject_string, ToString::to_string)
            );
        } else {
            let mut pyproject_file = File::create(&pyproject_path).unwrap();

            pyproject_file
                .write_all(updated_pyproject_string.as_bytes())
                .unwrap();

            if !keep_old_metadata {
                self.delete_pipenv_references().unwrap();
            }

            if !dry_run
                && !skip_lock
                && lock_dependencies(self.project_path.as_ref(), false).is_err()
            {
                warn!(
                    "An error occurred when locking dependencies, so \"{}\" was not created.",
                    "uv.lock".bold()
                );
            }

            if !ignore_locked_versions {
                let mut pyproject_updater = PyprojectUpdater {
                    pyproject: &mut updated_pyproject_string.parse::<DocumentMut>().unwrap(),
                };
                if let Some(updated_pyproject) = pyproject_updater.remove_constraint_dependencies()
                {
                    let mut pyproject_file = File::create(pyproject_path).unwrap();
                    pyproject_file
                        .write_all(updated_pyproject.to_string().as_bytes())
                        .unwrap();

                    // Lock dependencies a second time, to remove constraints from lock file.
                    if !dry_run
                        && !skip_lock
                        && lock_dependencies(self.project_path.as_ref(), true).is_err()
                    {
                        warn!("An error occurred when locking dependencies after removing constraints.");
                    }
                }
            }

            info!(
                "{}",
                "Successfully migrated project from Pipenv to uv!\n"
                    .bold()
                    .green()
            );
        }
    }

    #[cfg(test)]
    fn as_any(&self) -> &dyn Any {
        self
    }
}

impl Pipenv {
    fn perform_migration(
        &self,
        ignore_locked_versions: bool,
        dependency_groups_strategy: DependencyGroupsStrategy,
    ) -> String {
        let pipfile_content = fs::read_to_string(self.project_path.join("Pipfile")).unwrap();
        let pipfile: Pipfile = toml::from_str(pipfile_content.as_str()).unwrap();

        let mut uv_source_index: IndexMap<String, SourceContainer> = IndexMap::new();
        let (dependency_groups, uv_default_groups) =
            dependencies::get_dependency_groups_and_default_groups(
                &pipfile,
                &mut uv_source_index,
                dependency_groups_strategy,
            );

        let project = Project {
            // "name" is required by uv.
            name: Some(String::new()),
            // "version" is required by uv.
            version: Some("0.0.1".to_string()),
            requires_python: project::get_requires_python(pipfile.requires),
            dependencies: dependencies::get(pipfile.packages.as_ref(), &mut uv_source_index),
            ..Default::default()
        };

        let uv = Uv {
            package: Some(false),
            index: sources::get_indexes(pipfile.source),
            sources: if uv_source_index.is_empty() {
                None
            } else {
                Some(uv_source_index)
            },
            default_groups: uv_default_groups,
            constraint_dependencies: get_constraint_dependencies(
                ignore_locked_versions,
                &self.project_path.join("Pipfile.lock"),
            ),
        };

        let pyproject_toml_content =
            fs::read_to_string(self.project_path.join("pyproject.toml")).unwrap_or_default();
        let mut updated_pyproject = pyproject_toml_content.parse::<DocumentMut>().unwrap();
        let mut pyproject_updater = PyprojectUpdater {
            pyproject: &mut updated_pyproject,
        };

        pyproject_updater.insert_pep_621(&project);
        pyproject_updater.insert_dependency_groups(dependency_groups.as_ref());
        pyproject_updater.insert_uv(&uv);

        let mut visitor = PyprojectPrettyFormatter {
            parent_keys: Vec::new(),
        };
        visitor.visit_document_mut(&mut updated_pyproject);

        updated_pyproject.to_string()
    }

    fn delete_pipenv_references(&self) -> std::io::Result<()> {
        for file in FILES_TO_DELETE {
            let path = self.project_path.join(file);

            if path.exists() {
                remove_file(path)?;
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_perform_migration_python_full_version() {
        let tmp_dir = tempdir().unwrap();
        let project_path = tmp_dir.path();

        let pipfile_content = r#"
        [requires]
        python_full_version = "3.13.1"
        "#;

        let mut pipfile_file = File::create(project_path.join("Pipfile")).unwrap();
        pipfile_file.write_all(pipfile_content.as_bytes()).unwrap();

        let pipenv = Pipenv {
            project_path: PathBuf::from(project_path),
        };

        insta::assert_snapshot!(pipenv.perform_migration(true, DependencyGroupsStrategy::SetDefaultGroups), @r###"
        [project]
        name = ""
        version = "0.0.1"
        requires-python = "==3.13.1"

        [tool.uv]
        package = false
        "###);
    }

    #[test]
    fn test_perform_migration_empty_requires() {
        let tmp_dir = tempdir().unwrap();
        let project_path = tmp_dir.path();

        let pipfile_content = "[requires]";

        let mut pipfile_file = File::create(project_path.join("Pipfile")).unwrap();
        pipfile_file.write_all(pipfile_content.as_bytes()).unwrap();

        let pipenv = Pipenv {
            project_path: PathBuf::from(project_path),
        };

        insta::assert_snapshot!(pipenv.perform_migration(true, DependencyGroupsStrategy::SetDefaultGroups), @r###"
        [project]
        name = ""
        version = "0.0.1"

        [tool.uv]
        package = false
        "###);
    }
}
