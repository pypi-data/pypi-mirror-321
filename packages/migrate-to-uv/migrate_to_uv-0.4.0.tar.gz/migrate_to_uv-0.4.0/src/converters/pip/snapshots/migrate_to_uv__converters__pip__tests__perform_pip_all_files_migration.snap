---
source: src/converters/pip/mod.rs
expression: pip.perform_migration(true)
---
'''
[project]
name = ""
version = "0.0.1"
dependencies = [
    "anyio==4.7.0",
    "arrow==1.3.0",
    "certifi==2024.12.14",
    "click==8.1.8",
    "h11==0.14.0",
    "httpcore==1.0.7",
    "httpx==0.28.1",
    "idna==3.10",
    "markdown-it-py==3.0.0",
    "mdurl==0.1.2",
    "pygments==2.18.0",
    "python-dateutil==2.9.0.post0",
    "rich==13.9.4",
    "six==1.17.0",
    "sniffio==1.3.1",
    "types-python-dateutil==2.9.0.20241206",
    "uvicorn @ git+https://github.com/encode/uvicorn",
    "zstandard==0.23.0",
]

[dependency-groups]
dev = [
    "iniconfig==2.0.0",
    "packaging==24.2",
    "pluggy==1.5.0",
    "pytest==8.3.4",
    "ruff==0.8.4",
    "attrs==24.3.0",
    "mypy==1.14.1",
    "mypy-extensions==1.0.0",
    "referencing==0.35.1",
    "rpds-py==0.22.3",
    "types-jsonschema==4.23.0.20241208",
    "typing-extensions==4.12.2",
]

[tool.uv]
package = false
'''
